### PR TITLE
Enable implicit args and context bound implicits in @free traits

### LINKED ISSUE #365: context bounds not allowed in algebras
I had this algebra that worked in freestyle 0.2

```scala
  final case class Response[T](header: HttpResponseHeader, body: T)

  @free trait JsonHttpClient {
    def get[B: Decoder](uri: Uri,
                        headers: List[HttpHeader] = Nil): FS[Response[B]]

    // using application/x-www-form-urlencoded
    def postUrlencoded[A: UrlEncoded, B: Decoder](
      uri: Uri,
      payload: A,
      headers: List[HttpHeader] = Nil
    ): FS[Response[B]]
  }
```

the idea being that it can be a simple JSON HTTP client that takes in a circe `Decoder` for responses and responds as domain objects... so the caller can say what they expect and get it.

However, in freestyle 0.3 this fails with a macro error regarding `Decoder` implicit divergence (which I think is a red herring). Converting the context bound notation to use an explicit `implicit` parameter block, I get this warning

```
case classes should have a non-implicit parameter list; adapting to 'case class PostUrlencodedOP()(...)'
```

followed by this error

```
[error] <macro>:27: no arguments allowed for nullary method apply
```

which looks like implicits are just not being handled.

It's very annoying that `scala.meta` doesn't output the location of where the macro is being applied in terms of the file that generated it, because it can be super hard to localise where problems are coming from. @xeno-by can anything be done about that?

By explicit implicits I mean

```scala
  @free trait JsonHttpClient {
    def get[B](
      uri: Uri,
      headers: List[HttpHeader] = Nil
    )(
      implicit decoder: Decoder[B]
    ): FS[Response[B]]

    // using application/x-www-form-urlencoded
    def postUrlencoded[A, B](
      uri: Uri,
      payload: A,
      headers: List[HttpHeader] = Nil
    )(implicit encoder: UrlEncoded[A], decoder: Decoder[B]): FS[Response[B]]
  }
```

### PR DESCRIPTION
**Pull request by raulraja**
Fixes #365 /cc @fommil

### CODE REVIEW

**Review by andyscott — APPROVED** [review]
LGTM

**Comment by fommil** [comment]
Nice.

File: freestyle/shared/src/main/scala/freestyle/internal/free.scala [hunk 1]
```diff
@@ -201,7 +201,21 @@ private[internal] class Request(reqDef: Decl.Def, indexValue: Int) {
...
+    cbound <- tps.cbounds
+  } yield Term.Param(Nil, Term.fresh(cboundPrefix), Some(Type.Apply(cbound, Seq(Type.Name(tps.name.value)))), None)
+
+  val tparams = reqDef.tparams.map {
+    case p if p.cbounds.nonEmpty => p.copy(cbounds = Nil)
+    case p => p
+  }
+
+  val params: Seq[Term.Param] = reqDef.paramss.flatten.map {
+    case p @ param"..$mods $paramname: $atpeopt = $expropt" /* if p.hasMod(mod"implicit")*/ =>
```
Comment (fommil) [inline freestyle/shared/src/main/scala/freestyle/internal/free.scala:215]: I recall Eugene showing me a different way to check for mods. But I can remember, sorry.

Comment (raulraja) [reply freestyle/shared/src/main/scala/freestyle/internal/free.scala:215]: There is `scalameta-contrib` lib that brings the `hasMod` method but it has also some transitive dependencies we don't want in the core at this point.

**Pull request merged**